In [2]:
import copy
import pandas as pd
import numpy as np
import Levenshtein
from tqdm import tqdm

In [3]:
mf=pd.read_csv('../../../../paper_data_v3/irat_enet/results/animal_predictions/animal_combined_with_irat.csv')#.assign(is_irat=0)
mf=mf.rename(columns={'name':'id','ha_sequence':'ha','na_sequence':'na','emergence':'emergence_risk'})
mf.to_csv('v3_animalrisk.csv')

In [4]:
mf

id subtype        date  \
0                          A/Hong Kong/125/2017    H7N9  2017-05-01   
1                            A/Shanghai/02/2013    H7N9  2016-04-01   
2                          A/California/62/2018    H1N2  2019-07-01   
3                             A/Indiana/08/2011    H3N2  2012-12-01   
4                          A/Sichuan/06681/2021    H5N6  2021-10-01   
...                                         ...     ...         ...   
6372                  A/chicken/China/2106/2021    H9N2  2021-11-30   
6373                  A/chicken/China/2034/2020    H9N2  2020-01-20   
6374                  A/chicken/China/2040/2020    H9N2  2020-01-10   
6375  A/green-winged-teal/Georgia/DT-22246/2020    H5N8  2020-09-29   
6376            A/chicken/China/YCDC-KM213/2022    H9N2  2022-01-01   

     ha_accession                                                 ha  \
0       EPI977395  MNTQILVFALIAIIPTNADKICLGHHAVSNGTKVNTLTERGVEVVN...   
1       EPI448936  MNTQILVFALIAIIPTNADKICLGHHAVSNGTKVNTLTERGVEVVN...   
2      EPI1311361  MKVKLMVLLCTFTATYADTICVGYHANNSTDTVDTVLEKNVTVTHS...   
3       EPI344405  MKTIIAFSCILCLIFAQKLPGSDNSMATLCLGHHAVPNGTLVKTIT...   
4      EPI1883261  MENIVLLLAIVSLVKSDQICIGYHANNSTEQVDTIMEKNVTVTHAQ...   
...           ...                                                ...   
6372   EPI2116853  METASLITVLLVVTVSNADKICIGYQSTNSTETVDTLTENNVPVTH...   
6373   EPI2116765  METVSLITILLVAAVSNADKICIGYQSTNSTETVDTLTENNVPVTH...   
6374   EPI2116781  METVSLITILLVAAVSNADKICIGYQSTNSTETVDTLTENNVPVTH...   
6375   EPI1924318  MENIVLLLAIVSLVKSDQICIGYHANNSTEQVDTIMEKNVTVTHAQ...   
6376   EPI2915506  METVSLITILLAATVSNADKICIGYQSTNSTETVDTLTENNVPVTH...   

     na_accession                                                 na  \
0       EPI977394  MNPNQKILCTSATAITIGAIAVLIGIANLGLNIGLHLKPGCNCSHS...   
1       EPI448938  MNPNQKILCTSATAIIIGAIAVLIGMANLGLNIGLHLKPGCNCSHS...   
2      EPI1311360  MNPNQKIITIGSISLTLAAMCFLMQTAILVTNVTLHFNQCECHYPP...   
3       EPI344404  MNPNQKIITIGSVSLIIATICFLMQIAILVTTVTLHFKQHDYNSPP...   
4      EPI1883263  MNPNQKITCISATGVTLSIVSLLIGITNLGLNIGLHYKVSDSTTIN...   
...           ...                                                ...   
6372   EPI2116851  MNPNQKIIAIGSVSLTIAIICFLMQIAILTTTMTLHFRQNECSNPS...   
6373   EPI2116763  MNPNQKIIAIGSVSLTIAIICFLMQIAILTTTMTLHFRQNECSNPS...   
6374   EPI2116779  MNPNQKIIAIGSVSLTIAIICFLMQIAILTTTMTLHFRQNECSNPS...   
6375   EPI1924320  MNPNQKIATIGSISLGLVVFNVLLHALSIILMVLALGKSENNGICK...   
6376   EPI2915504  MNPNQKITAIGSVSLIIAIICLLMQIAILTTTMTLHFEHKECSTPS...   

       ha_risk   na_risk  geom_mean_risk  emergence_risk  emergence_low  \
0     0.000010  0.000010        0.000010        7.732746       6.187663   
1     0.000010  0.000010        0.000010        7.732746       6.187663   
2     0.000010  0.000010        0.000010        7.732746       6.187663   
3     0.000055  0.000080        0.000066        7.636444       6.104862   
4     0.000010  0.194167        0.001393        6.529923       5.153469   
...        ...       ...             ...             ...            ...   
6372  0.151614  0.045829        0.083357        3.053955       2.164810   
6373  0.152940  0.045829        0.083720        3.049986       2.161398   
6374  0.152942  0.045829        0.083721        3.049980       2.161392   
6375  0.248596  0.028458        0.084110        3.045752       2.157757   
6376  0.207073  0.039085        0.089963        2.984417       2.105021   

      emergence_high  is_irat  emergence_score  impact_score  \
0           9.179121        1              6.5           7.5   
1           9.179121        1              6.4           7.2   
2           9.179121        1              5.8           5.7   
3           9.073535        1              6.0           4.5   
4           7.860356        1              5.3           6.3   
...              ...      ...              ...           ...   
6372        4.049335        0              NaN           NaN   
6373        4.044984        0              NaN   

In [9]:
# parameters
threshold=6.25  # draw tree above this emergenec risk threshold
CONSTRUCT_PHYLO=True  # use constructed tree (tree construction takes hours)
OUTPUT_DIR='./'
PHYLO_TREE_DIR='./'
PHYLO_DIR='./'
#COMBINED_RESULTS='../../../../paper_data/irat_enet/results/animal_predictions/combined_results.csv'
COMBINED_RESULTS='v3_animalrisk.csv'
#CONSTRUCT_PHYLO=False
num_collapsed=19  # number of mutations within which leaves are collapsed
VERBOSE=False

# class definition to hold multi sequence data

In [15]:
class SeqInfo(object):
    """Holds information regarding the sequence.
    
    """
    def __init__(self, seq, 
                 protein,
                 accession,
                 subtype=None,
                 id=None,
                 name=None,
                 host=None, 
                 date=None, 
                 erisk=None,
                 irisk=None,
                 risk_flag=None,
                 country=None):
        self.name = name
        self.id = id
        self.protein=protein
        self.subtype=subtype        
        self.seq = seq
        self.accession = accession 
        self.host = host
        self.date = date
        self.erisk = erisk
        self.irisk = irisk
        self.risk_flag = risk_flag
        self.country = country
        
class MultipleSeqInfo(object):
    """Holds information regarding multiple sequences.
    
    Args:
        dataframe (pandas.DataFrame): list of records parsed from NCBI
        accessionname (str): column name for accession id
        proteinname (str): protein name 
        risk_threshold (float): emergence risk threshold to compute distance matrix
    """
    def __init__(self,
                 dataframe,
                 accessionname,
                 proteinname,
                 risk_threshold=6.2):
        
        self.seq_infos = {}
        self.risk_threshold = risk_threshold
        for i in np.arange(dataframe.index.size):
            record=dataframe.iloc[i,:]
            seqinfo = SeqInfo(
                name=record.id,
                seq=record[proteinname], 
                protein=proteinname,
                accession=record[accessionname],
                subtype=record.subtype,
                erisk=record.emergence_risk,
                risk_flag = record.emergence_risk > self.risk_threshold,
                host=None,
                date=None,
                country=None)
            #print(record.predicted_emergence_score > self.risk_threshold)
            self.seq_infos[seqinfo.accession] = seqinfo
            
    
    def compute_L_distance_matrix(self):
        highriskseq = pd.DataFrame.from_dict({key:val.seq 
                                              for (key,val) in self.seq_infos.items() 
                                              if val.risk_flag},orient='index',columns=['seq'])
        num=highriskseq.index.size
        d=np.zeros([num,num])
        for x in tqdm(np.arange(num*num)):
            j=x//num
            i=x-num*j
            if i > j:
                d[i,j] = Levenshtein.distance(highriskseq.seq.values[i],
                                                  highriskseq.seq.values[j])
        ds=pd.DataFrame(d)        
        ds=(ds+ds.transpose())
        ds.columns=highriskseq.index.values
        self.highriskdistancematrix=ds.copy()
        
        self.highriskdistancematrix.to_csv('dm24_'+str(self.risk_threshold)+'.csv',index=None)
        return 
    
    
    def accessions_to_subtype(self, accessions):
        """Create a dictionary mapping the accession to the host.
        """
        
        subtypes = []
        for accession in accessions:
            seqinfo = self.seq_infos[accession]
            subtypes.append(seqinfo.subtype)
            
        return subtypes

    def accessions_to_host(self, accessions):
        """Create a dictionary mapping the accession to the host.
        """
        
        hosts = []
        for accession in accessions:
            seqinfo = self.seq_infos[accession]
            hosts.append(seqinfo.host)
        return hosts
           
    

In [16]:
df=pd.read_csv(COMBINED_RESULTS,index_col=0).reset_index()
#ALLinfoHA=MultipleSeqInfo(df.reset_index(),'ha_accession','ha',risk_threshold=6.054)
ALLinfoHA=MultipleSeqInfo(df.reset_index(),'ha_accession','ha',risk_threshold=threshold)

In [17]:
df[df.emergence_risk>threshold].subtype.value_counts()

H5N1    1732
H3N2     930
H1N1     481
H1N2     418
H3N8      14
H3N1       7
H7N9       2
H3N3       2
H5N6       1
H9N2       1
H5N2       1
H3N6       1
Name: subtype, dtype: int64

# generate distance matrix

In [18]:
ALLinfoHA.compute_L_distance_matrix()

100%|████████████████████████████| 12880921/12880921 [03:43<00:00, 57665.11it/s]


# tree construction from distance matrix using biopython

In [19]:
from Bio.Phylo import TreeConstruction
from Bio import Phylo
from Bio.SeqRecord import SeqRecord
from Bio.Seq import Seq
from Bio.Align import MultipleSeqAlignment
from Bio import Entrez
from Bio import SeqIO


def load_dm(file_, upper_diag=True):
    """Load the distance matrix. 
    
    Also, do some preprocessing. 
    """
    
    df = pd.read_csv(file_)
    #df.set_index('Unnamed: 0', inplace=True)
    #assert np.all(df.columns == df.index)
    
    # drop duplicate columns after reading csv
    #df = df.loc[:, ~df.columns.str.replace("(\.\d+)$", "").duplicated()]
    
    if upper_diag:
        df = df + df.T
    return df

def save_tree(tree, file_name, save_type='xml'):
    """Saved the created phylogenetic tree."""
    
    if save_type == 'pickle':
        graph = Phylo.to_networkx(tree)
        save_pickled(graph, file_name)
    elif save_type == 'xml':
        Phylo.write(tree, file_name, 'phyloxml')
    else:
        raise ValueError('Not a correct save type.')
    
def pandas_dm_to_biopython_dm(dm):
    """Convert the pandas distance matrix to the biopython distance matrix.
    
    Returns:
        biopython distance matrix
    """
    
    accessions = dm.columns
    bio_dm = []
    for i, accession in enumerate(accessions):
        bio_dm.append(list(dm.iloc[i, :i+1].values))
        
    bio_dm = TreeConstruction._DistanceMatrix(
        list(dm.columns), 
        bio_dm)
    
    return bio_dm

def distance_matrix_to_phylo_tree(dm, outfile=None):
    """Create a phylogenetic tree from the distance matrix."""
    
    dm = pandas_dm_to_biopython_dm(dm)
    
    treeConstructor = TreeConstruction.DistanceTreeConstructor()
    tree = treeConstructor.nj(dm)
    
    if outfile is not None:
        save_tree(tree, outfile)

# ete3 function, not all of these are used here

In [20]:
from ete3 import Tree, TreeStyle
from ete3 import Phyloxml
from ete3 import AttrFace, faces, Tree, NodeStyle, TreeStyle

def load_pickled(file_name):
    with open(file_name, 'rb') as f:
        return pickle.load(f, encoding='latin')


def get_farthest_node(tree, sequence):
    return (tree&sequence).get_farthest_node()

def get_all_accessions_from_tree(tree):
    return [leaf_node.name for leaf_node in tree.get_leaves()]

def remove_certain_hosts_from_tree(tree, hosts):
    """Remove leaf nodes if the host of that leaf is in `hosts`"""
    
    tree = copy.deepcopy(tree)
    
    removed_accessions = []
    for leaf_node in tree.get_leaves():
        if leaf_node.host in hosts:
            leaf_node.detach()
            
    return tree

def set_midpoint_outgroup(tree):
    tree.set_outgroup(tree.get_midpoint_outgroup())


def load_tree(filename, type_='phyloxml'):
    """Load saved phylogenetic tree.
    """
    
    if type_ == 'phyloxml':
        project = Phyloxml()
        project.build_from_file(filename)

        for tree in project.get_phylogeny():
            break

        t=tree
        
    elif type_ == 'newick':
        t = Tree(filename, format=1)
    else:
        raise ValueError('Not a correct type.')
    
    return t

# construct tree from distance matrix

In [21]:
#CONSTRUCT_PHYLO=True
if CONSTRUCT_PHYLO:
    ALL_dm_ldistance = load_dm(
        OUTPUT_DIR + 'dm24_'+str(threshold)+'.csv', 
        upper_diag=False)
    
    distance_matrix_to_phylo_tree(
        ALL_dm_ldistance, PHYLO_TREE_DIR + 'ldistance'+str(threshold)+'.xml')

# convert phyloxml tree to newick tree to manipulate trees

In [22]:

Phylo.convert(
    PHYLO_DIR + 'ldistance'+str(threshold)+'.xml','phyloxml',
    PHYLO_DIR + 'ldistance'+str(threshold)+'.nhx','newick')

ltree = load_tree(
    PHYLO_DIR + 'ldistance'+str(threshold)+'.nhx',
    type_='newick')

# label nodes in tree to add other attributes like subtype risk etc

In [23]:
def bandify(val,min=6.05,max=6.71):
    maptoten=int(np.ceil(((val-min)/(max-min))*10))
    return ' '+u'\u2580'*maptoten

def label_nodes(
        tree, 
        recordinfo):
    """Label the nodes of the tree.
    
    We label nodes on whether:
        it is covid19
    """
    
    tree = copy.deepcopy(tree)
    
    for node in tree:
        name = node.name      
        node.subtype = recordinfo.seq_infos[name].subtype
        node.erisk =recordinfo.seq_infos[name].erisk
        node.id = recordinfo.seq_infos[name].name + bandify(recordinfo.seq_infos[name].erisk,min=threshold)
        if VERBOSE:
            print(node.name,node.subtype,node.id,node.erisk)
    return tree

# construct labelled tree

In [24]:
labelled_tree=label_nodes(
    ltree, ALLinfoHA)

# functions to collapse simialr leaves

In [25]:
def prune_nodes(t):
    # collapsed nodes are labeled, so you locate them and prune them
    for n in t.search_nodes(collapsed=True):
        for ch in n.get_children():
            ch.detach()
            
            
def mean(array):
    return sum(array)/float(len(array))

def cache_distances(tree):
    ''' precalculate distances of all nodes to the root''' 
    node2rootdist = {tree:0}
    for node in tree.iter_descendants('preorder'):
        node2rootdist[node] = node.dist + node2rootdist[node.up]
    return node2rootdist

def closest_node(node, node2tips, root_distance):
    """Find the closest node."""
    
    tips = []
    distances = []
    for tip in node2tips[node]:
        distances.append(root_distance[tip]-root_distance[node])
        tips.append(tip)
        #     index = np.argmin([root_distance[tip]-root_distance[node] for tip in node2tips[node]])
    index = np.argmin(distances)
    return tips[index]

def riskiest_node(node, node2tips):
    """Find the closest node."""
    
    tips = []
    risks = []
    for tip in node2tips[node]:
        risks.append(tip.erisk)
        tips.append(tip)
        #     index = np.argmin([root_distance[tip]-root_distance[node] for tip in node2tips[node]])
    index = np.argmax(risks)
    return tips[index]

def collapse(tree, min_dist,AllrecordInfo):
    # cache the tip content of each node to reduce the number of times the tree is traversed
    
    tree = copy.deepcopy(tree)
    
    node2tips = tree.get_cached_content()
    root_distance = cache_distances(tree)

    for node in tree.get_descendants('preorder'):
        if not node.is_leaf():
            avg_distance_to_tips = mean([root_distance[tip]-root_distance[node]
                                         for tip in node2tips[node]])
            if VERBOSE:
                print(avg_distance_to_tips)
            if avg_distance_to_tips < min_dist:
                # do whatever, ete support node annotation, deletion, labeling, etc.
            
                #closest_name = closest_node(node, node2tips, root_distance).name
                closest_name = riskiest_node(node, node2tips).name
                node.subtype = AllrecordInfo.seq_infos[closest_name].subtype
                node.id = AllrecordInfo.seq_infos[closest_name].name + bandify(AllrecordInfo.seq_infos[closest_name].erisk,min=threshold)
                node.name = '%s (%g)' %(closest_name,avg_distance_to_tips)
                
            
                node.add_features(collapsed=True)

                # set drawing attribute so they look collapsed when displayed with tree.show()
                node.img_style['draw_descendants'] = False

    return tree

In [26]:
# collapse leaved

In [27]:
num_collapsed=19
ltree_collapsed = collapse(
    labelled_tree, 
    min_dist=num_collapsed, 
    AllrecordInfo=ALLinfoHA)

prune_nodes(ltree_collapsed)

# code for actual rendering

In [38]:
# COLBAT='DarkRed'
# COLRAT='SteelBlue'
COLH3N2='Blue'
COLH1N1='DarkRed'
COLH1N2='FF1111'
COLH5N1='#55FFFF'
COLH7N9='DarkGreen'
COLH9N2='#BD890F'
COLDEF='BLACK'

FS=50
PW=10


def nodeAttribConstruct(color, node):
    N = AttrFace(
        "id", fsize=FS, 
        text_prefix=" ",penwidth=PW,ftype='Arial',
        fgcolor=color,fstyle='bold')
    faces.add_face_to_node(N, node, 1, position="branch-right")
    return N

def layout(node):
    if node.is_leaf():
        if  node.subtype == 'H1N1':
            N = nodeAttribConstruct(COLH1N1,node)
        elif node.subtype == 'H3N2':
            N = nodeAttribConstruct(COLH3N2,node)
        elif node.subtype == 'H7N9':
            N = nodeAttribConstruct(COLH7N9,node)
        elif node.subtype == 'H9N2':
            N = nodeAttribConstruct(COLH9N2,node)
        elif node.subtype == 'H1N2':
            N = nodeAttribConstruct(COLH1N2,node)
        elif node.subtype == 'H5N1':
            N = nodeAttribConstruct(COLH5N1,node)
        else:
            N = nodeAttribConstruct(COLDEF,node)
            

            
def render_tree(tree, outfile):# all_seq_data, display_type='nearest_host'):
    """Render the tree inside the file to a circular 
    phylogenetic tree.
    
    NOTE: outfile should be in .pdf for best visuals
    Returns:
    """
    #tree = Tree(nwfile,format=1)

    ts = TreeStyle()
    ns = NodeStyle()
    ts.show_leaf_name = False
    #ts.rotation = 90
    ts.mode = "r"
    #ts.arc_start = -360 # 0 degrees = 3 o'clock
    #ts.arc_span = 360
    ts.scale=5
    ts.show_scale=False
    ts.branch_vertical_margin = .5 # 10 pixels between adjacent branches
    # ts.show_branch_length=True
    #ts.min_leaf_separation=10
    #ts.optimal_scale_level='full'
    #ts.branch_vertical_margin=0
    
    ns.hz_line_width=2
    ns.vt_line_width=1
    #ts.layout_fn = layout
    ns["vt_line_width"] = 16
    ns["hz_line_width"] = 16
    #     ns['fsize'] = 20
    for n in tree.traverse():
        n.set_style(ns)
        
    #all_accessions = all_seq_data['accessions'].values
    for n in tree:
        ts.layout_fn = layout

        
    tree.set_style(ns)
    tree.set_style(ts)
    
    ax=tree.render(
        outfile, 
        dpi=300, 
        w=500,
        tree_style=ts)

In [39]:
r=5
medrisknames=[]
for node in ltree_collapsed:
    if ALLinfoHA.seq_infos[node.name.split()[0]].erisk > r:
        medrisknames=np.append(medrisknames,
                                ALLinfoHA.seq_infos[node.name.split()[0]].name)
medrisknames=df[df.id.isin(medrisknames)][['id','subtype',
                                               'ha_accession',
                                               'na_accession',
                                               'emergence_risk']].sort_values('emergence_risk',
                                                                                         ascending=False)
medrisknames.to_csv('allriskystrains_collapsed.csv',index=None)
medrisknames[medrisknames.id.str.contains('Sha')]

Empty DataFrame
Columns: [id, subtype, ha_accession, na_accession, emergence_risk]
Index: []

# Render phylogenetic trees

In [40]:
ax=render_tree(
    labelled_tree,
    './riskyphylo'+str(threshold)+'_irat.pdf')

In [41]:
ax=render_tree(
    ltree_collapsed,
    'riskyphylo'+str(threshold)+'_collapsed_'+str(num_collapsed)+'_irat.pdf')

In [33]:
ltree_collapsed

Tree node 'Inner3587' (0x7f3c2271b7f)

# Find high risk strains which are on distinct branches on phylogenetic tree

In [31]:
r=6.057
r=6.25
highrisknames=[]
for node in ltree_collapsed:
    if ALLinfoHA.seq_infos[node.name.split()[0]].erisk > r:
        highrisknames=np.append(highrisknames,
                                ALLinfoHA.seq_infos[node.name.split()[0]].name)
highrisknamesdf=df[df.id.isin(highrisknames)][['id','subtype',
                                               'ha_accession',
                                               'na_accession',
                                               'impact_risk',
                                               'emergence_risk']].sort_values('emergence_risk',
                                                                                         ascending=False)
highrisknamesdf = highrisknamesdf.rename(columns={'id':'strain',
                                                  'ha_accession':'HA accession',
                                                  'na_accession':'NA accession',
                                                  'impact_risk':'predicted IRAT impact',
                                                  'emergence_risk':'predicted IRAT emergence'}).set_index('strain')
highrisknamesdf#.drop_duplicates()

subtype HA accession NA accession  \
strain                                                                    
A/swine/Missouri/A02524711/2020          H1N1   EPI1818121   EPI1818122   
A/swine/Indiana/A02524710/2020           H3N2   EPI1818137   EPI1818138   
A/swine/North_Carolina/A02479173/2020    H1N1   EPI1780425   EPI1780426   
A/Camel/Inner_Mongolia/XL/2020           H7N9   EPI2026200   EPI2026202   
A/swine/Tennessee/A02524414/2022         H1N1   EPI2149257   EPI2149258   
...                                       ...          ...          ...   
A/swine/Illinois/A02525253/2021          H3N2   EPI1910375   EPI1910376   
A/swine/Kansas/A02245381/2020(H1N1)      H1N1   EPI1777723   EPI1777724   
A/canine/Texas/21-011409-001/2021        H3N2   EPI1896555   EPI1896557   
A/swine/Iowa/A02246996/2021              H1N1   EPI2146133   EPI2146134   
A/mink/China/chick_embryo/2020           H9N2   EPI2161544   EPI2161548   

                                       predicted IRAT impact  \
strain                                                         
A/swine/Missouri/A02524711/2020                     6.767254   
A/swine/Indiana/A02524710/2020                      6.720489   
A/swine/North_Carolina/A02479173/2020               6.713621   
A/Camel/Inner_Mongolia/XL/2020                      6.699016   
A/swine/Tennessee/A02524414/2022                    6.650119   
...                                                      ...   
A/swine/Illinois/A02525253/2021                     6.397379   
A/swine/Kansas/A02245381/2020(H1N1)                 6.344661   
A/canine/Texas/21-011409-001/2021                   6.344011   
A/swine/Iowa/A02246996/2021                         6.326870   
A/mink/China/chick_embryo/2020                      6.318778   

                                       predicted IRAT emergence  
strain                                                           
A/swine/Missouri/A02524711/2020                        6.782248  
A/swine/Indiana/A02524710/2020                         6.729258  
A/swine/North_Carolina/A02479173/2020                  6.721477  
A/Camel/Inner_Mongolia/XL/2020                         6.704867  
A/swine/Tennessee/A02524414/2022                       6.649429  
...                                                         ...  
A/swine/Illinois/A02525253/2021                        6.361542  
A/swine/Kansas/A02245381/2020(H1N1)                    6.301257  
A/canine/Texas/21-011409-001/2021                      6.300499  
A/swine/Iowa/A02246996/2021                            6.280888  
A/mink/China/chick_embryo/2020                         6.271620  

[68 rows x 5 columns]

In [32]:
COLDICT={'H1N1':'Red3!20','H3N2':'Blue1!30','H7N9':'Green3!50','H9N2':'DarkOrange!40'}
def rowcolor(row):
    return '\\rowcolor{' + COLDICT[row.subtype]+'}' + row['strain']

highrisknamesdf1 = highrisknamesdf.reset_index(drop=False)
highrisknamesdf1['strain']=highrisknamesdf1.apply(rowcolor,axis=1)
highrisknamesdf1=highrisknamesdf1.set_index('strain')
highrisknamesdf1

subtype HA accession  \
strain                                                                    
\rowcolor{Red3!20}A/swine/Missouri/A02524711/2020     H1N1   EPI1818121   
\rowcolor{Blue1!30}A/swine/Indiana/A02524710/2020     H3N2   EPI1818137   
\rowcolor{Red3!20}A/swine/North_Carolina/A02479...    H1N1   EPI1780425   
\rowcolor{Green3!50}A/Camel/Inner_Mongolia/XL/2020    H7N9   EPI2026200   
\rowcolor{Red3!20}A/swine/Tennessee/A02524414/2022    H1N1   EPI2149257   
...                                                    ...          ...   
\rowcolor{Blue1!30}A/swine/Illinois/A02525253/2021    H3N2   EPI1910375   
\rowcolor{Red3!20}A/swine/Kansas/A02245381/2020...    H1N1   EPI1777723   
\rowcolor{Blue1!30}A/canine/Texas/21-011409-001...    H3N2   EPI1896555   
\rowcolor{Red3!20}A/swine/Iowa/A02246996/2021         H1N1   EPI2146133   
\rowcolor{DarkOrange!40}A/mink/China/chick_embr...    H9N2   EPI2161544   

                                                   NA accession  \
strain                                                            
\rowcolor{Red3!20}A/swine/Missouri/A02524711/2020    EPI1818122   
\rowcolor{Blue1!30}A/swine/Indiana/A02524710/2020    EPI1818138   
\rowcolor{Red3!20}A/swine/North_Carolina/A02479...   EPI1780426   
\rowcolor{Green3!50}A/Camel/Inner_Mongolia/XL/2020   EPI2026202   
\rowcolor{Red3!20}A/swine/Tennessee/A02524414/2022   EPI2149258   
...                                                         ...   
\rowcolor{Blue1!30}A/swine/Illinois/A02525253/2021   EPI1910376   
\rowcolor{Red3!20}A/swine/Kansas/A02245381/2020...   EPI1777724   
\rowcolor{Blue1!30}A/canine/Texas/21-011409-001...   EPI1896557   
\rowcolor{Red3!20}A/swine/Iowa/A02246996/2021        EPI2146134   
\rowcolor{DarkOrange!40}A/mink/China/chick_embr...   EPI2161548   

                                                    predicted IRAT impact  \
strain                                                                      
\rowcolor{Red3!20}A/swine/Missouri/A02524711/2020                6.767254   
\rowcolor{Blue1!30}A/swine/Indiana/A02524710/2020                6.720489   
\rowcolor{Red3!20}A/swine/North_Carolina/A02479...               6.713621   
\rowcolor{Green3!50}A/Camel/Inner_Mongolia/XL/2020               6.699016   
\rowcolor{Red3!20}A/swine/Tennessee/A02524414/2022               6.650119   
...                                                                   ...   
\rowcolor{Blue1!30}A/swine/Illinois/A02525253/2021               6.397379   
\rowcolor{Red3!20}A/swine/Kansas/A02245381/2020...               6.344661   
\rowcolor{Blue1!30}A/canine/Texas/21-011409-001...               6.344011   
\rowcolor{Red3!20}A/swine/Iowa/A02246996/2021                    6.326870   
\rowcolor{DarkOrange!40}A/mink/China/chick_embr...               6.318778   

                                                    predicted IRAT emergence  
strain                                                                        
\rowcolor{Red3!20}A/swine/Missouri/A02524711/2020                   6.782248  
\rowcolor{Blue1!30}A/swine/Indiana/A02524710/2020                   6.729258  
\rowcolor{Red3!20}A/swine/North_Carolina/A02479...                  6.721477  
\rowcolor{Green3!50}A/Camel/Inner_Mongolia/XL/2020                  6.704867  
\rowcolor{Red3!20}A/swine/Tennessee/A02524414/2022                  6.649429  
...                                                                      ...  
\rowcolor{Blue1!30}A/swine/Illinois/A02525253/2021                  6.361542  
\rowcolor{Red3!20}A/swine/Kansas/A02245381/2020...                  6.301257  
\rowcolor{Blue1!30}A/canine/Texas/21-011409-001...                  6.300499  
\rowcolor{Red3!20}A/swine/Iowa/A02246996/2021                       6.280888  
\rowcolor{DarkOrange!40}A/mink/China/chick_embr...                  6.271620  

[68 rows x 5 columns]

In [33]:
from zedstat.textable import textable
textable(highrisknamesdf1,#pd.concat([highrisknamesdf1.head(30),highrisknamesdf1.tail(10)]),
         tabname='highrisk35.tex',
         FORMAT='%1.4f',INDEX=True,
         TABFORMAT='L{1.95in}|L{.25in}|L{.60in}|L{.6in}|C{1in}|C{1in}',LNTERM='\\\\\n')
! cat highrisk35.tex

\begin{tabular}{L{1.95in}|L{.25in}|L{.60in}|L{.6in}|C{1in}|C{1in}}\hline
strain&subtype& HA  accession & NA  accession & predicted  IRAT  impact & predicted  IRAT  emergence \\
 \rowcolor{Red3!20}A/swine/Missouri/A02524711/2020 &H1N1&EPI1818121&EPI1818122&6.7673&6.7822\\
 \rowcolor{Blue1!30}A/swine/Indiana/A02524710/2020 &H3N2&EPI1818137&EPI1818138&6.7205&6.7293\\
 \rowcolor{Red3!20}A/swine/North\_Carolina/A02479173/2020 &H1N1&EPI1780425&EPI1780426&6.7136&6.7215\\
 \rowcolor{Green3!50}A/Camel/Inner\_Mongolia/XL/2020 &H7N9&EPI2026200&EPI2026202&6.6990&6.7049\\
 \rowcolor{Red3!20}A/swine/Tennessee/A02524414/2022 &H1N1&EPI2149257&EPI2149258&6.6501&6.6494\\
 \rowcolor{Red3!20}A/swine/Minnesota/A02635976/2021 &H1N1&EPI1912208&EPI1912209&6.5776&6.5670\\
 \rowcolor{Blue1!30}A/swine/Chile/VN1401-5054/2020 &H3N2&EPI1974975&EPI1974978&6.5318&6.5149\\
 \rowcolor{Blue1!30}A/swine/Italy/56910/2020 &H3N2&EPI2142217&EPI2142173&6.5292&6.5119\\
 \rowcolor{Blue1!30}A/swine/Minnesota/A02245643/2020 &H3N2

In [34]:
count=0
Subtype={'H1N1':0,'H3N2':0,'H7N9':0,'H9N2':0}
MaxriskStrain={'H1N1':None,'H3N2':None,'H7N9':None,'H9N2':None}
Subtype_strat={ 6.0:{'H1N1':0,'H3N2':0,'H7N9':0,'H9N2':0},  6.2:{'H1N1':0,'H3N2':0,'H7N9':0,'H9N2':0},              6.3:{'H1N1':0,'H3N2':0,'H7N9':0,'H9N2':0},
               6.4:{'H1N1':0,'H3N2':0,'H7N9':0,'H9N2':0},6.5:{'H1N1':0,'H3N2':0,'H7N9':0,'H9N2':0}}
for node in ltree_collapsed:
    Subtype[node.subtype]=Subtype[node.subtype]+1
    if MaxriskStrain[node.subtype] is None:
        MaxriskStrain[node.subtype]=(ALLinfoHA.seq_infos[node.name.split()[0]].name,
                                     ALLinfoHA.seq_infos[node.name.split()[0]].erisk)
    else:
        if MaxriskStrain[node.subtype][1]<ALLinfoHA.seq_infos[node.name.split()[0]].erisk:
            MaxriskStrain[node.subtype]=(ALLinfoHA.seq_infos[node.name.split()[0]].name,
                                         ALLinfoHA.seq_infos[node.name.split()[0]].erisk)
    for r in [6.0, 6.2,6.3,6.4,6.5]:
        if r<ALLinfoHA.seq_infos[node.name.split()[0]].erisk:
            #print(node.subtype,ALLinfoHA.seq_infos[node.name.split()[0]].erisk)
            Subtype_strat[r][node.subtype]=Subtype_strat[r][node.subtype]+1
        
    count=count+1
maxriskdf = pd.DataFrame(MaxriskStrain)
Subtype_strat_df = pd.DataFrame(Subtype_strat)
display(maxriskdf)
display(Subtype_strat_df)
print(count)
df[df.id.isin(maxriskdf.iloc[0,:].values)]

H1N1                            H3N2  \
0  A/swine/Missouri/A02524711/2020  A/swine/Indiana/A02524710/2020   
1                         6.782248                        6.729258   

                             H7N9                            H9N2  
0  A/Camel/Inner_Mongolia/XL/2020  A/mink/China/chick_embryo/2020  
1                        6.704867                         6.27162

6.0  6.2  6.3  6.4  6.5
H1N1   61   57   54   11    4
H3N2   11   11   11    6    3
H7N9    1    1    1    1    1
H9N2    1    1    0    0    0

75


id subtype ha_accession na_accession  \
218   A/swine/Missouri/A02524711/2020    H1N1   EPI1818121   EPI1818122   
1902   A/swine/Indiana/A02524710/2020    H3N2   EPI1818137   EPI1818138   
5814   A/Camel/Inner_Mongolia/XL/2020    H7N9   EPI2026200   EPI2026202   
6049   A/mink/China/chick_embryo/2020    H9N2   EPI2161544   EPI2161548   

                                                     ha  \
218   MKAILVVMLYTFTTANADTLCIGYHANNSTDTVDTVLEKNVTVTHS...   
1902  MKTIIALSYILCLVFAQKIPGNDNGTATLCLGHHAVPNGTIVKTIT...   
5814  MNTQILVFALIAIIPTNADKICLGHHAVSNGTKVNTLTEREVEVVN...   
6049  METVSLITILLAATVSNADKICIGYQSSNSTETVDTLTENNVPVTH...   

                                                     na  \
218   MNPNQKIITIGSICMTIGTANLILQIGNIISIWVSHSIQIGNQSQI...   
1902  MNPNQKIITIGSVSLTISTICFFMQIAILITTIALHFKQYEFNSPP...   
5814  MNPNQKILCTSATAITIGAIAVLIGIANLGLNIGLHLKPGCNCSHS...   
6049  MNPNQKITAIGSVSLIIAIICLLMQIAILTMTMTLHFGQKECSNPS...   

                                                ha_risk  \
218   [0.007702821053332326, 0.00794167832708008, 0....   
1902  [0.008423100311281287, 0.0044835948774065975, ...   
5814  [0.025061424930009856, 0.01711340116551936, 0....   
6049  [0.016907786181883842, 0.016907786181883842, 0...   

                                                na_risk  \
218   [0.0016000691389129699, 0.0019043683598783635,...   
1902  [0.006596476585994605, 0.006619358238135353, 0...   
5814  [0.0011227960579890436, 0.0012230115569459082,...   
6049  [0.04614861940588559, 0.04614861940588559, 0.0...   

                                    geometric_mean_risk  emergence_risk  \
218   [0.0035107045233124007, 0.003888943420830784, ...        6.782248   
1902  [0.00745404480700585, 0.00544779961894918, 0.0...        6.729258   
5814  [0.0053046082908169015, 0.004574919387714034, ...        6.704867   
6049  [0.027933331156592276, 0.027933331156592276, 0...        6.271620   

      emergence_risk_var  impact_risk  impact_risk_var  
218             0.000006     6.767254         0.000005  
1902            0.000613     6.720489         0.000476  
5814            0.004832     6.699016         0.003755  
6049            0.000000     6.318778         0.000000

In [35]:
from zedstat.textable import textable
Subtype_strat_df.columns=[str(x) for x in Subtype_strat_df.columns]
Subtype_strat_df.index.name='subtype'
Subtype_strat_df=Subtype_strat_df.reset_index()
textable(Subtype_strat_df,#pd.concat([highrisknamesdf1.head(30),highrisknamesdf1.tail(10)]),
         tabname='riskycount.tex',
         FORMAT='%1.4f',INDEX=False,
         TABFORMAT='L{.75in}|L{.25in}|L{.60in}|L{.6in}|L{.6in}|L{.6in}',LNTERM='\\\\\n')
! cat riskycount.tex

\begin{tabular}{L{.75in}|L{.25in}|L{.60in}|L{.6in}|L{.6in}|L{.6in}}\hline
subtype&6.0&6.2&6.3&6.4&6.5\\
H1N1&61&57&54&11&4\\
H3N2&11&11&11&6&3\\
H7N9&1&1&1&1&1\\
H9N2&1&1&0&0&0\\
\hline\end{tabular}


In [36]:
def write_fasta(seqs, fasta_file, wrap=80):
    """Write sequences to a fasta file.

    Parameters
    ----------
    seqs : dict[seq_id] -> seq
        Sequences indexed by sequence id.
    fasta_file : str
        Path to write the sequences to.
    wrap: int
        Number of AA/NT before the line is wrapped.
    """
    with open(fasta_file, 'w') as f:
        for gid, gseq in seqs.items():
            f.write('>{}\n'.format(gid))
            for i in range(0, len(gseq), wrap):
                f.write('{}\n'.format(gseq[i:i + wrap])) 

#for i,name in zip(df[df.id.isin(maxriskdf.iloc[0,:].values)].ha.values,
#                 df[df.id.isin(maxriskdf.iloc[0,:].values)].ha_accession.values):
#    write_fasta({name:i},name+'.fasta')
    

In [37]:
id_list=['A/swine/North_Carolina/A02479173/2020','A/swine/Tennessee/A02524414/2022']
RFILES={name+'('+subtype+')':i for i,name,subtype in zip(df[df.id.isin(id_list)].ha.values,
                 df[df.id.isin(id_list)].ha_accession.values,
                 df[df.id.isin(id_list)].subtype.values)}

In [38]:
FASTANAME='risky4.fasta'
FASTAFILES={name+'('+subtype+')':i for i,name,subtype in zip(df[df.id.isin(maxriskdf.iloc[0,:].values)].ha.values,
                 df[df.id.isin(maxriskdf.iloc[0,:].values)].ha_accession.values,
                 df[df.id.isin(maxriskdf.iloc[0,:].values)].subtype.values)}

In [43]:
human_dom_seq='../../../qnet_predictions/results/dominant_sequences_2021_2022.csv'
domf=pd.read_csv(human_dom_seq,index_col=0)
domf=domf.assign(subtype=[x.split('_')[1].upper() for x in domf.index.values],protein=[x.split('_')[2].upper() for x in domf.index.values])
domf

name  \
strain                                         
north_h1n1_ha           A/Ireland/20935/2022   
north_h1n1_na      A/Cote_d'Ivoire/3729/2021   
north_h3n2_ha  A/Human/New_York/PV60641/2022   
north_h3n2_na            A/Stockholm/10/2022   
south_h1n1_ha             A/Abidjan/457/2021   
south_h1n1_na      A/Cote_D'Ivoire/1496/2021   
south_h3n2_ha               A/Darwin/9a/2021   
south_h3n2_na     A/India/PUN-NIV301718/2021   

                                                        sequence subtype  \
strain                                                                     
north_h1n1_ha  MKAILVVLLYTFTTANADTLCIGYHANNSTDTVDTVLEKNVTVTHS...    H1N1   
north_h1n1_na  MNPNQKIITIGSICMAIGTANLILQIGNIISIWVSHSIQIGNQSQI...    H1N1   
north_h3n2_ha  MKTIIALSNILCLVFAQKIPGNDNSTATLCLGHHAVPNGTIVKTIT...    H3N2   
north_h3n2_na  MNPNQKIITIGSVSLTISTICFFMQIAILITTVTLHFKQYEFNSPP...    H3N2   
south_h1n1_ha  MKAILVVLLYTFTTANADTLCIGYHANNSTDTVDTVLEKNVTVTHS...    H1N1   
south_h1n1_na  MNPNQKIITIGSICMAIGTANLILQIGNIISIWVSHSIQIGNQSQI...    H1N1   
south_h3n2_ha  MKTIIALSNILCLVFAQKIPGNDNSTATLCLGHHAVPNGTIVKTIT...    H3N2   
south_h3n2_na  MNPNQKIITIGSVSLTISTICFFMQIAILITTVTLHFKQYEFNSPP...    H3N2   

              protein  
strain                 
north_h1n1_ha      HA  
north_h1n1_na      NA  
north_h3n2_ha      HA  
north_h3n2_na      NA  
south_h1n1_ha      HA  
south_h1n1_na      NA  
south_h3n2_ha      HA  
south_h3n2_na      NA

In [55]:
human_dom_seq='../../../../paper_data/enet_predictions/results/dominant_sequences/north_h1n1_ha/north_h1n1_ha_21_22.csv'
domf=pd.read_csv(human_dom_seq,index_col=0)
domf=domf.assign(subtype='H1N1',protein='HA')
domf=domf[domf.protein=='HA'].head(2)
DOMFILES={'DOM_HUMAN'+'('+subtype+')':i for i,name,subtype in zip(domf.sequence.values,
                 domf.id.values,
                 domf.subtype.values)}
DOMFILES.update(FASTAFILES)
DOMFILES

{'DOM_HUMAN(H1N1)': 'MKAILVVLLYTFTTANADTLCIGYHANNSTDTVDTVLEKNVTVTHSVNLLEDKHNGKLCKLRGVAPLHLGKCNIAGWILGNPECESLSTARSWSYIVETSNSDNGTCYPGDFINYEELREQLSSVSSFERFEIFPKTSSWPNHDSDKGVTAACPHAGAKSFYKNLIWLVKKGNSYPKLNQTYINDKGKEVLVLWGIHHPPTIAAQESLYQNADAYVFVGTSRYSKKFKPEIATRPKVRDQEGRMNYYWTLVEPGDKITFEATGNLVVPRYAFTMERDAGSGIIISDTPVHDCNTTCQTPEGAINTSLPFQNVHPITIGKCPKYVKSTKLRLATGLRNVPSIQSRGLFGAIAGFIEGGWTGMVDGWYGYHHQNEQGSGYAADLKSTQNAIDKITNKVNSVIEKMNTQFTAVGKEFNHLEKRIENLNKKVDDGFLDIWTYNAELLVLLENERTLDYHDSNVKNLYEKVRNQLKNNAKEIGNGCFEFYHKCDNTCMESVKNGTYDYPKYSEEAKLNREKIDGVKLESTRIYQILAIYSTVASSLVLVVSLGAISFWMCSNGSLQCRIC',
 'EPI1818121(H1N1)': 'MKAILVVMLYTFTTANADTLCIGYHANNSTDTVDTVLEKNVTVTHSVNLLEDKHNGKLCKLRGVAPLHLGKCNIAGWILGNPECESLSTARSWSYIVETSNSDNGTCYPGDFINYEELREQLSSVSSFERFEIFPKTSSWPNHDSDNGVTAACPHAGAKSFYKNLIWLVKKGKSYPKINQTYINDKGKEVLVLWGIHHPPTIADQQSLYQNADAYVFVGTSRYSKKFKPEIATRPKVRDQEGRMNYYWTLVEPGDKITFEATGNLVAPRYAFTMERDAGSGIIISDTPVHDCNTTCQTPEGAINTSLPFQNVHPITIGKCPKYVKSTKLRLATGLRNVPSIQSRGLFGAIAGFIEGGWTGMVDGWYGYHHQNEQGSGYAADLKSTQNAI

In [56]:
DOMFILES.update(RFILES)

In [57]:
DOMFILES

{'DOM_HUMAN(H1N1)': 'MKAILVVLLYTFTTANADTLCIGYHANNSTDTVDTVLEKNVTVTHSVNLLEDKHNGKLCKLRGVAPLHLGKCNIAGWILGNPECESLSTARSWSYIVETSNSDNGTCYPGDFINYEELREQLSSVSSFERFEIFPKTSSWPNHDSDKGVTAACPHAGAKSFYKNLIWLVKKGNSYPKLNQTYINDKGKEVLVLWGIHHPPTIAAQESLYQNADAYVFVGTSRYSKKFKPEIATRPKVRDQEGRMNYYWTLVEPGDKITFEATGNLVVPRYAFTMERDAGSGIIISDTPVHDCNTTCQTPEGAINTSLPFQNVHPITIGKCPKYVKSTKLRLATGLRNVPSIQSRGLFGAIAGFIEGGWTGMVDGWYGYHHQNEQGSGYAADLKSTQNAIDKITNKVNSVIEKMNTQFTAVGKEFNHLEKRIENLNKKVDDGFLDIWTYNAELLVLLENERTLDYHDSNVKNLYEKVRNQLKNNAKEIGNGCFEFYHKCDNTCMESVKNGTYDYPKYSEEAKLNREKIDGVKLESTRIYQILAIYSTVASSLVLVVSLGAISFWMCSNGSLQCRIC',
 'EPI1818121(H1N1)': 'MKAILVVMLYTFTTANADTLCIGYHANNSTDTVDTVLEKNVTVTHSVNLLEDKHNGKLCKLRGVAPLHLGKCNIAGWILGNPECESLSTARSWSYIVETSNSDNGTCYPGDFINYEELREQLSSVSSFERFEIFPKTSSWPNHDSDNGVTAACPHAGAKSFYKNLIWLVKKGKSYPKINQTYINDKGKEVLVLWGIHHPPTIADQQSLYQNADAYVFVGTSRYSKKFKPEIATRPKVRDQEGRMNYYWTLVEPGDKITFEATGNLVAPRYAFTMERDAGSGIIISDTPVHDCNTTCQTPEGAINTSLPFQNVHPITIGKCPKYVKSTKLRLATGLRNVPSIQSRGLFGAIAGFIEGGWTGMVDGWYGYHHQNEQGSGYAADLKSTQNAI

In [58]:
write_fasta(DOMFILES,FASTANAME)

In [49]:
from Bio import SeqIO
from Bio.Blast import NCBIWWW
my_query = SeqIO.read("EPI1818121.fasta", format="fasta")
result_handle = NCBIWWW.qblast("blastp", "nr", my_query.seq)
blast_result = open("my_blast.xml", "w")
blast_result.write(result_handle.read())
blast_result.close()
result_handle.close()

In [50]:
def parse_blast(resultfile): #takes in the BLAST result, outputs list that can be made into csv
    from Bio.Blast import NCBIXML
    result_handle = open(resultfile)
    blast_records = NCBIXML.parse(result_handle)
    csv_list = []
    
    header = [  'accession','Query',
                'Name', 'Length', 'Score', 'Expect',
                'QueryStart', 'QueryEnd',
                'SubjectStart', 'SubjectEnd','pct'
            ]
    
    #csv_list.append(header)
    count = 0
    for blast_record in blast_records:
        '''help(blast_record.alignments[0].hsps[0])''' # these give help info for the parts 
        '''help(blast_record.alignments[0])        '''
        count +=1
        
        query = blast_record.query
        for alignment in blast_record.alignments:

            name = alignment.title
            length = alignment.length
    
            hsp = alignment.hsps[0] # I don't know if we will ever have more than one, so might as well take the first one.
            score = hsp.score
            expect = hsp.expect
            querystart = hsp.query_start
            queryend = hsp.query_end
            subjectstart = hsp.sbjct_start
            subjectend = hsp.sbjct_end
            pct=hsp.positives/hsp.align_length
            accession=alignment.accession
            row = [accession,query,name,length,score,expect,querystart,queryend,subjectstart,subjectend,pct]
            csv_list.append(row)
            
    result_handle.close()
    return pd.DataFrame(csv_list,columns=header)
 

In [51]:
csv_list=parse_blast('my_blast.xml')
csv_list

accession                    Query  \
0   QIA58660  unnamed protein product   
1   WHU33766  unnamed protein product   
2   QKG28282  unnamed protein product   
3   QIZ32536  unnamed protein product   
4   QJI52882  unnamed protein product   
5   QJT24758  unnamed protein product   
6   QKV49724  unnamed protein product   
7   QIQ45861  unnamed protein product   
8   QKM75781  unnamed protein product   
9   QJX57625  unnamed protein product   
10  WBG95744  unnamed protein product   
11  QIQ45830  unnamed protein product   
12  QOY44156  unnamed protein product   
13  QIQ45216  unnamed protein product   
14  QMX87502  unnamed protein product   
15  QJD23117  unnamed protein product   
16  QIQ45677  unnamed protein product   
17  QJT24395  unnamed protein product   
18  WBG95731  unnamed protein product   
19  QIQ45346  unnamed protein product   
20  QIZ32395  unnamed protein product   
21  QIZ32042  unnamed protein product   
22  QIZ33251  unnamed protein product   
23  QIQ44996  unnamed protein product   
24  QTJ06578  unnamed protein product   
25  WBG95844  unnamed protein product   
26  QJT24499  unnamed protein product   
27  QIV14744  unnamed protein product   
28  QIQ45654  unnamed protein product   
29  QIA57747  unnamed protein product   
30  QIZ31801  unnamed protein product   
31  QKM75836  unnamed protein product   
32  QKS74061  unnamed protein product   
33  QIH95333  unnamed protein product   
34  QIZ32602  unnamed protein product   
35  QJT24351  unnamed protein product   
36  QLV95699  unnamed protein product   
37  QKM76085  unnamed protein product   
38  QJI52636  unnamed protein product   
39  QIQ45093  unnamed protein product   
40  QJT24428  unnamed protein product   
41  QIA58044  unnamed protein product   
42  QIQ45082  unnamed protein product   
43  QJT24466  unnamed protein product   
44  QIZ31721  unnamed protein product   
45  QTK18161  unnamed protein product   
46  QIV14634  unnamed protein product   
47  QJI52715  unnamed protein product   
48  QTK18105  unnamed protein product   
49  QIZ32778  unnamed protein product   

                                                 Name  Length   Score  Expect  \
0   gb|QIA58660.1| hemagglutinin [Influenza A viru...     566  2969.0     0.0   
1   gb|WHU33766.1| hemagglutinin [Influenza A viru...     566  2967.0     0.0   
2    gb|QKG28282.1| hemagglutinin [Influenza A virus]     566  2967.0     0.0   
3    gb|QIZ32536.1| hemagglutinin [Influenza A virus]     566  2967.0     0.0   
4    gb|QJI52882.1| hemagglutinin [Influenza A virus]     566  2967.0     0.0   
5    gb|QJT24758.1| hemagglutinin [Influenza A virus]     566  2967.0     0.0   
6   gb|QKV49724.1| hemagglutinin [Influenza A viru...     566  2967.0     0.0   
7    gb|QIQ45861.1| hemagglutinin [Influenza A virus]     566  2967.0     0.0   
8    gb|QKM75781.1| hemagglutinin [Influenza A virus]     566  2966.0     0.0   
9    gb|QJX57625.1| hemagglutinin [Influenza A virus]     566  2966.0     0.0   
10   gb|WBG95744.1| hemagglutinin [Influenza A virus]     566  2966.0     0.0   
11   gb|QIQ45830.1| hemagglutinin [Influenza A virus]     566  2966.0     0.0   
12  gb|QOY44156.1| hemagglutinin [Influenza A viru...     566  2966.0     0.0   
13  gb|QIQ45216.1| hemagglutinin [Influenza A viru...     566  2965.0     0.0   
14  gb|QMX87502.1| hemagglutinin [Influenza A viru...     566  2965.0     0.0   
15   gb|QJD23117.1| hemagglutinin [Influenza A virus]     566  2965.0     0.0   
16  gb|QIQ45677.1| hemagglutinin [Influenza A viru...     566  2965.0     0.0   
17  gb|QJT24395.1| hemagglutinin [Influenza A viru...     566  2965.0     0.0   
18   gb|WBG95731.1| hemagglutinin [Influenza A virus]     566  2965.0     0.0   
19   gb|QIQ45346.1| hemagglutinin [Influenza A virus]     566  2965.0     0.0   
20  gb|QIZ32395.1| hemagglutinin [Influenza A viru...     566  2964.0     0.0   
21   gb|QIZ32042.1| hemagglutinin [Influenza A virus]     566  2964.0     0.0   
22   gb|QIZ33251.1| hemagglutinin [Influenza A v

In [52]:
from Bio import Entrez
Entrez.email = "ishanu@uchicago.edu"     # Always tell NCBI who you are
handle = Entrez.esearch(db="protein", term="QJT24340")
record = Entrez.read(handle)
id=record["IdList"]
id

['1841676337']

In [53]:
handle = Entrez.efetch(db="protein", rettype="gb", retmode="text", id="1834373412")
#record = SeqIO.read(handle, "genbank")
handle.read()

'LOCUS       QJD23117                 566 aa            linear   VRL 28-APR-2020\nDEFINITION  hemagglutinin [Influenza A virus].\nACCESSION   QJD23117\nVERSION     QJD23117.1\nDBSOURCE    accession MT372532.1\nKEYWORDS    .\nSOURCE      Influenza A virus\n  ORGANISM  Influenza A virus\n            Viruses; Riboviria; Orthornavirae; Negarnaviricota;\n            Polyploviricotina; Insthoviricetes; Articulavirales;\n            Orthomyxoviridae; Alphainfluenzavirus; Alphainfluenzavirus\n            influenzae.\nREFERENCE   1  (residues 1 to 566)\n  AUTHORS   Swine Surveillance,A.\n  TITLE     USDA NAHLN Voluntary Swine Influenza Surveillance\n  JOURNAL   Unpublished\nREFERENCE   2  (residues 1 to 566)\n  AUTHORS   Swine Surveillance,A.\n  TITLE     Direct Submission\n  JOURNAL   Submitted (23-APR-2020) DiVL, USDA NAHLN Voluntary Swine Influenza\n            Surveillance, 1920 Dayton Ave., Ames, IA 50010, USA\nCOMMENT     Method: conceptual translation.\nFEATURES             Location/Qual

H9N2: minks are probable mixing vessels
https://www.ncbi.nlm.nih.gov/nuccore/ON870694.1
 https://www.tandfonline.com/doi/full/10.1080/22221751.2021.1899058
 Pandemic influenza, typically caused by the reassortment of human and avian influenza viruses, can result in severe or fatal infections in humans. Timely identification of potential pandemic viruses must be a priority in influenza virus surveillance. However, the range of host species responsible for the generation of novel pandemic influenza viruses remains unclear. In this study, we conducted serological surveys for avian and human influenza virus infections in farmed mink and determined the susceptibility of mink to prevailing avian and human virus subtypes. The results showed that farmed mink were commonly infected with human (H3N2 and H1N1/pdm) and avian (H7N9, H5N6, and H9N2) influenza A viruses. Correlational analysis indicated that transmission of human influenza viruses occurred from humans to mink, and that feed source was a probable route of avian influenza virus transmission to farmed mink. Animal experiments showed that mink were susceptible and permissive to circulating avian and human influenza viruses, and that human influenza viruses (H3N2 and H1N1/pdm), but not avian viruses, were capable of aerosol transmission among mink. These results indicate that farmed mink could be highly permissive “mixing vessels” for the reassortment of circulating human and avian influenza viruses. 
 
 H7N9
 https://www.bmj.com/content/347/bmj.f4752?tab=responses
 HH transmission has been suspected
 Asian lineage H7N9 virus is rated by the Influenza Risk Assessment Tool as having the greatest potential to cause a pandemic, as well as potentially posing the greatest risk to severely impact public health if it were to achieve sustained human-to-human transmission.
 
 H3N2 swine
 https://www.cdc.gov/flu/swineflu/spotlights/first-human-infection-2022.htm
 Human infection detected in US this Aug
 
 H1N1
 Submitted (03-NOV-2020) USDA Swine Surveillance, USDA Swine
            Surveillance, 1920 Dayton, Ames, IA 50010, USA
COMMENT     Method: conceptual translation.



In [5]:
mf.columns

Index(['id', 'subtype', 'date', 'ha_accession', 'ha', 'na_accession', 'na',
       'ha_risk', 'na_risk', 'geom_mean_risk', 'emergence_risk',
       'emergence_low', 'emergence_high', 'is_irat', 'emergence_score',
       'impact_score', 'mean_low_acceptable_emergence',
       'mean_high_acceptable_emergence', 'mean_low_acceptable_impact',
       'mean_high_acceptable_impact'],
      dtype='object')

In [9]:
mf[mf.ha_accession=='EPI2143526']

id subtype        date ha_accession  \
4635  A/swine/Liaoning/FS1084/2020    H1N1  2020-05-10   EPI2143526   

                                                     ha na_accession  \
4635  MEARLLVLFCAFTTLNADTICVGYHANNSTDTVDTILEKNVTVTHS...   EPI2143552   

                                                     na   ha_risk   na_risk  \
4635  MNPNQKIITIGSICMTIGITSLILQIGSIISIWISHSIQTKNQNQS...  0.004303  0.004471   

      geom_mean_risk  emergence_risk  emergence_low  emergence_high  is_irat  \
4635        0.004386        5.660569       4.405993        6.907204        0   

      emergence_score  impact_score  mean_low_acceptable_emergence  \
4635              NaN           NaN                            NaN   

      mean_high_acceptable_emergence  mean_low_acceptable_impact  \
4635                             NaN                         NaN   

      mean_high_acceptable_impact  
4635                          NaN

In [12]:
mf[mf.emergence_risk>7.7].reset_index()

index                                      id subtype        date  \
0       0                    A/Hong Kong/125/2017    H7N9  2017-05-01   
1       1                      A/Shanghai/02/2013    H7N9  2016-04-01   
2       2                    A/California/62/2018    H1N2  2019-07-01   
3      23   A/swine/North_Carolina/A02751333/2022    H3N2  2022-12-13   
4      24         A/swine/Missouri/A02524408/2023    H3N2  2023-01-13   
5      25            A/swine/Iowa/CEIRS-1495/2023    H1N2  2023-06-15   
6      26      A/swine/Tver_region/RII-81-1S/2023    H1N1  2023-02-17   
7      27          A/swine/Indiana/A01812310/2020    H1N1  2020-08-17   
8      28             A/swine/Ohio/OH20-2231/2020    H1N1  2020-02-04   
9      29             A/swine/Iowa/A02479281/2020    H1N1  2020-04-27   
10     30      A/swine/North_Carolina/O41356/2023    H3N2  2023-05-01   
11     31             A/swine/Iowa/A02524545/2020    H1N1  2020-08-14   
12     32       A/swine/Moscow_region/RII-47/2023    H1N1  2023-01-27   
13     33             A/swine/Iowa/A02479209/2020    H1N1  2020-04-08   
14     34         A/swine/Georgia/CEIRS-0147/2023    H1N1  2023-01-19   
15     35             A/swine/Iowa/A02479310/2020    H1N1  2020-05-07   
16     36             A/swine/Ohio/A02479097/2020    H1N1  2020-03-02   
17     37  A/swine/Metropolitana/VN1401-4881/2020    H1N1  2020-02-28   
18     38             A/swine/Iowa/A02524793/2020    H1N1  2020-10-26   
19     39             A/swine/Iowa/A02479425/2020    H1N1  2020-06-02   
20     40             A/swine/Iowa/A02861468/2023    H1N1  2023-05-02   
21     41  A/swine/Metropolitana/VN1401-4882/2020    H1N1  2020-02-28   
22     42         A/swine/Ohio/ISU-A02861688/2023    H1N1  2023-07-27   
23     43         A/swine/Georgia/CEIRS-0343/2023    H1N1  2023-01-19   
24     44         A/swine/Georgia/CEIRS-0290/2023    H1N1  2023-01-19   
25     45         A/swine/Iowa/ISU-A02861755/2023    H1N1  2023-08-29   
26     46         A/swine/Nebraska/A02245865/2021    H1N1  2021-01-05   
27     47   A/swine/North_Carolina/A02524922/2020    H1N1  2020-11-25   
28     48         A/swine/Iowa/ISU-A02862121/2023    H1N1  2023-12-11   
29     49         A/swine/Iowa/ISU-A02862159/2023    H1N1  2023-12-15   
30     50         A/swine/Missouri/A02758605/2023    H3N2  2023-02-07   
31     51     A/swine/Colorado/ISU-A02758606/2023    H3N2  2023-04-26   
32     52     A/swine/Colorado/ISU-A02758609/2023    H3N2  2023-05-11   
33     53     A/swine/South_Dakota/A02524617/2020    H1N1  2020-09-09   
34     54         A/swine/Missouri/A01104146/2020    H1N1  2020-12-14   

   ha_accession                                                 ha  \
0     EPI977395  MNTQILVFALIAIIPTNADKICLGHHAVSNGTKVNTLTERGVEVVN...   
1     EPI448936  MNTQILVFALIAIIPTNADKICLGHHAVSNGTKVNTLTERGVEVVN...   
2    EPI1311361  MKVKLMVLLCTFTATYADTICVGYHANNSTDTVDTVLEKNVTVTHS...   
3    EPI2396893  MKTIIALSNILCLVFAQKIPGNDNSTATLCLGHHAVPNGTIVKTIT...   
4    EPI2433963  MKTIIALSNILCLVFAQKIPGNDNSTATLCLGHHAVPNGTIVKTIT...   
5    EPI2908633  MKVKLMVLLCTFTATYADTICIGYHANNSTDTIDTVLEKNVTVTHS...   
6    EPI2965146  MKAILVVMLYTFTTANADTLCIGYHANNSTDTVDTVLEKNVTVTHS...   
7    EPI1833157  MKAILVVLLYTFTTANADTLCIGYHANNSTDTVDTVLEKNVTVTHS...   
8    EPI2497342  MKAILVVLLYTFTTANADTLCIGYHANNSTDTVDTVLEKNVTVTHS...   
9    EPI1775980  MKAILVVLLYTFTTANADTLCIGYHANNSTDTVDTVLEKNVTVTHS...   
10   EPI2718359  MKTIIALSNILCLVFAQKIPGNDNSTATLCLGHHAVPNGTIVKTIT...   
11   EPI1832802  MKAILVVLLYTFTTANADTLCIGYHANNSTDTVDTVLEKNVTVTHS...   
12   EPI2965139  MKAILVVMLYTFTTANADTLCIGYHANNSTDTVDTVLEKNVTVTHS...   
13   EPI1775888  MKAILVVLLYTFTTANADTLCIGYHANNSTDTVDTVLEKNVTVTHS...   
14   EPI2908514  MKAILVVMLYTFTTANADTLCIGYHANNSTDTVDTVLEKNVTVTHS...   
15   EPI1768609  MKAILVVLLYTFTTANADTLCIGYHANNSTDTVDTVLEKNVTVTHS...   
16   EPI1780120  MKAILVVLLYTFTTANADTLCIGYHANNSTDTVDTVLEKNVTVTHS...   
17   EPI1974925  MKAILVVLLYTFTTANADTLCIGYHANNSTDTVDTVLEKNVTVTHS...   
18   EPI1907263  MKAILVVLLYTFTTANADTLCIGYHANN